### PR TITLE
Change MouseTracker's interface for clarity. Simplify MouseRegion's implementation.

### PR DESCRIPTION
**Pull request by dkwingsmt**
## Description

This PR redesigns the interface between MouseTracker and RendererBinding&RenderView for 2 purposes:

* Allow MouseTracker to use the hit test result that GestureBinding has already done, instead of doing its own again.
* Make MouseTracker feel more "pure" by no longer secretly listening to the pointer router.

With the new interface, `MouseTracker`'s constructor receives no parameters. It must be notified by being called externally, provided with hit test result. In tests, the hit test is no longer mocked by rebuilding `MouseTracker`s, but by changing the child of the `renderView`, which is closer to "testing the behavior".

This PR also simplifies the structure of `RenderMouseRegion`. Since it no longer relies on layers, it doesn't need to propagate changes by triggering repaints. Nor does it need to turn itself inactivate when no mouse devices are connected.

This PR also extracts the common utility code between `mouse_tracker_test` and `mouse_tracker_cursor_test`, and make them prettier.

## Related Issues

- N/A

## Tests

- Some tests in `mouse_tracker_test` now also test hover events.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/rendering/binding.dart [hunk 1]
```diff
@@ -284,7 +296,24 @@ mixin RendererBinding on BindingBase, ServicesBinding, SchedulerBinding, Gesture
 
   void _handlePersistentFrameCallback(Duration timeStamp) {
     drawFrame();
-    _mouseTracker.schedulePostFrameCheck();
+    _scheduleMouseTrackerUpdate();
+  }
+
+  bool _debugMouseTrackerUpdateScheduled = false;
+  void _scheduleMouseTrackerUpdate() {
```
Comment (gspencergoog) [inline packages/flutter/lib/src/rendering/binding.dart:303]: Couldn't you leave `_mouseTracker.schedulePostFrameCheck()`, and just replace the body of that function with the contents of this one?

I.e., doesn't this code belong in the mouse tracker, and not the binding?

**Review by gspencergoog — COMMENTED** [review]
I'll bet it feels good to delete all that code. :-)

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

(modulo moving that function)

File: packages/flutter/lib/src/rendering/binding.dart (hunk 1 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/rendering/binding.dart:303]: I think the idea is that `MouseTracker` does not schedule for itself. It provides a method that should be scheduled, and the binding schedule things, just like how the binding subscribes to the pointer router. In this way, `MouseTracker` only provides 2 methods that is called outside, and is agnostic by whom, when, and how often they are called.

Comment (gspencergoog) [reply packages/flutter/lib/src/rendering/binding.dart:303]: OK, makes sense.

**Pull request merged**